ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/jca/TransactionImporterImple.java

```diff
@@ -122,6 +112,14 @@ public TransactionImple recoverTransaction(Uid actId)
 		if (recovered.baseXid() == null)
 		    throw new IllegalArgumentException();
 		
+		return addImportedTransaction(recovered);
+	}
+
+	private TransactionImple addImportedTransaction(TransactionImple importedTransaction) throws XAException {
+		SubordinateXidImple importedXid = new SubordinateXidImple(importedTransaction.baseXid());
+		TransactionImpleHolder holder = new TransactionImpleHolder();
+		TransactionImpleHolder prevHolder = _transactions.putIfAbsent(importedXid, holder);
+
```

Comment:
In practice, none of these exceptions apply to ConcurrentHashMap with this code.  Basically, they all relate to preconditions which are met; you don't have to worry about environmental factors causing sporadic behavior changes in this case.